packages/ui-components/README.md

```diff
@@ -43,3 +43,47 @@ The icons are organized into various categories in `./style/icons`, based on whe
   - launcher
   - statusbar
   - settingeditor
+
+## Notes on design of JLIcon
+
+Design goals for JLIcon (already implemented):
+
+- the one true icon
+  - create a single, canonical, simple (as possible) way to set up and use icons in jlab
+- every icon is a symbol
+  - each icon is defined exactly once, and is thereafter referred to via a unique symbol
+  - to use an icon `fooIcon` outside of the file in which it is defined, the pattern will be to import `fooIcon`
+  - this enables compile-time checking; helps to ensure that icons are specified correctly when used (as opposed to the old pattern of specifying icon via a `string` with name or className)
+- every icon is flexible
+  - can used in any context in which icons are used in jlab
+    - every icon can be passed into Lumino
+    - every icon can create a DOM node
+    - every icon can create a React component
+- dynamic lookup (for when absolutely needed)
+  - Use dynamic lookup for the few cases in which an icon can only be specified as a string (such as in json schema files)
+  - In all other cases, force (or at least strongly encourage) the pattern of import-icon
+- reusable
+  - every defined icon can be used any number of times in any set of contexts
+  - where an icon is defined should not matter; all icons defined in core and extensions are reusable in any other extension
+
+Design goals for JLIcon (partially implemented):
+
+- remove all other ways of creating icons (though leave an escape hatch)
+  - need to deprecate, then later remove, `iconClass` from a number of interfaces
+- replacable
+  - all icons can be customized by replacing their svg dynamically during runtime
+    - currently, I'm leaning towards the idea that icon replacements should be an (optional) part of a theme
```

Comment:
That sounds good! Previously, some icons used to be in the themes right? https://github.com/jupyterlab/jupyterlab/blob/master/packages/theme-light-extension/style/urls.css#L6